sdk/go/pulumi/rpc.go

```diff
@@ -795,11 +797,17 @@ func unmarshalOutput(ctx *Context, v resource.PropertyValue, dest reflect.Value)
 		if err != nil {
 			return false, err
 		}
-		resV := reflect.ValueOf(res).Elem()
-		if !resV.Type().AssignableTo(dest.Type()) {
+		resV := reflect.ValueOf(res)
+		if !allocatedPointer && resV.Kind() == reflect.Ptr && dest.Type().Kind() == reflect.Interface &&
+			resV.Elem().Type().AssignableTo(dest.Type()) {
+			dest.Set(resV)
+			return secret, nil
+		}
```

Comment:
Not sure if this block warrants a comment explaining the "why" that the PR description alludes to/if there's a one-line example that captures why this block is necessary/important?

Reply:
Added a comment :heavy_check_mark: